components/user_profile/user_profile.tsx

```diff
@@ -54,6 +60,20 @@ export default class UserProfile extends PureComponent<UserProfileProps> {
         this.overlay = ref;
     }
 
+    generateColor = (username: string, background: string): string => {
+        let userColor = background;
+        let userAndSalt = username;
+        const checker = new ContrastChecker();
+        const colorHash = new ColorHash();
+
+        while (!checker.isLevelCustom(userColor, background, 4.5)) {
```

Comment:
I'm concerned about the time added to the `render()` function by this `while` loop.
From what I can tell (correct me if i'm wrong) we are basically just rolling the dice over and over using the color hash and adding strings as we go until we get a color that we deem acceptable.

This feels like on a larger scale (think like 1000+ different users in one channel for example) that we might end up introducing some poor performance by the `user_profile` component.

I'm wondering can we add some options to the `ColorHash` object (see here: https://github.com/zenozeng/color-hash#custom-hue) that dictate what the HSV values are bound within? That way we could ahead of time calculate what HSV values would the minimum/maximum required to provide adequate contrast? It would at least mitigate the amount of loops that the code go through before it finds a colour.

cc @hmhealey in case he has any ideas.

Reply:
Yep, I considered this. I spent a lot of time playing around with the lightness constraints that ColorHash offers. Unfortunately lightness of the color is not the same as relative luminance, and this led to basically indistinguishably-dark or -light versions of the colors (think extreme pastels or earth tones.)

One mitigating factor is that users don't affect each other -- eg calculating the color constant for each user is dependent only on themselves and the background color. So there's not going to be quadratic explosion in the number of users, it's linear. 

I could try strategies like caching the background color + user so that it's only calculated once per theme refresh. I ran an experiment on a few thousand randomly generated usernames and the default background colors and I think it took at most ~4 iterations to find an acceptable color. I can go find the data from that if it helps.

Reply:
Thanks for the insight @3ach.
I think that its promising that it likely won't take many iterations to find the correct colour, and I definitely like the idea of caching this information so that we don't update it unless absolutely necessary, that's certainly a step in the right direction.

I'm still a little concerned what the impact might be on lower-end systems with thousands of users. I would like to see what the impact would be with some CPU throttling added with a few thousand users in the channel. Fortunately, it's unlikely we'll have 1000+ users rendered all at once since the center channel is designed not to render too many post components.

Reply:
Okay, sure, I'm happy to add the caching. I have no idea how I could run a test with CPU throttling... is there a guide or other accepted practice you are aware of?

Reply:
I believe there's a way you can do it in Developer Tools on Chrome (there's a CPU throttling dropdown somewhere, I forget exactly where). As for measurements, I think just a profile comparison (before and after caching) might yield some insights for us.

Reply:
Perhaps we could add a cap on that where we just give up and accept that a colour is a duplicate after a certain number of retries? We've done something similar in a few other places where something could theoretically take a very long time in rare cases, and while that case shouldn't come up in practice, it's nice to know it'll always perform at least okay

Reply:
@devinbinnie - I can go ahead and do a perf test as you described.
@hmhealey - The iterations are to ensure high contrast between background and a given username, not differentiation between usernames. If it's all right to give up on the contrast after a certain number of iterations, I'm happy to add that in.